### ISSUE TITLE
Support tags on sequences and mappings

### ISSUE THREAD

**Issue by ExpHP** [dtolnay/serde-yaml#147]
See https://users.rust-lang.org/t/help-using-serde-yaml-to-deserialize-value-with-very-many-possible-types/37058

```rust
#[derive(Debug, Deserialize)]
enum Parameter {
    vec2(ArrayVec<[f32; 2]>),
    vec3(ArrayVec<[f32; 3]>),
    vec4(ArrayVec<[f32; 4]>),
    curve(Vec<CurveVal>),
    str32(String),
    str64(String),
    str256(String),
    string(String),
}
```

Currently, this enum type is capable of using tags in place of singleton mappings to disambiguate between its String variants:

```yaml
- !str32 "hello"
- !string "world"
```

But it cannot use tags for the sequence-type variants: (it can only use singleton mappings)

```yaml
# These don't work
- !vec3 [1.0, 0.5, 0.3]
- !curve [1.0, 0.5, 0.3]

# Required workaround
- vec3: [1.0, 0.5, 0.3]
- curve: [1.0, 0.5, 0.3]
```

A solution to this would appear to be blocked on https://github.com/chyh1990/yaml-rust/issues/147, as the upstream parsing library currently does not emit tags in its `SequenceStart` or `MappingStart` events.

---

**Aside:** I find it rather amusing how both of these issues ended up being number 147. 😄

**Comment by aloucks** [comment]
It seems that tags don't work on numbers either.

Modified from `test_de.rs`:

```rust
#[test]
fn test_enum_tag() {
    #[derive(Deserialize, PartialEq, Debug)]
    enum E {
        A(String),
        B(String),
        C(f32), // <-- this is new
    }
    #[derive(Deserialize, PartialEq, Debug)]
    struct Data {
        a: E,
        b: E,
        c: E
    }
    let yaml = unindent(
        "
        ---
        a: !A foo
        b: !B bar
        c: !C 1.0",
    );
    let expected = Data {
        a: E::A("foo".into()),
        b: E::B("bar".into()),
        c: E::C(1.0),
    };
    test_de(&yaml, &expected);
}
```

```
thread 'test_enum_tag' panicked at 'called `Result::unwrap()` on an `Err` value: Message("invalid type: string \"1.0\", expected f32", Some(Pos { marker: Marker { index: 30, line: 4, col: 6 }, path: "c.C" }))', src\libcore\result.rs:1165:5
```

**Comment by dtolnay** [comment]
Fixed in 0.9.

**Comment by pcrumley** [comment]
Is there a supported way such that the former workaround of using e.g. `vec3: ...` instead of `!vec3` continues to work, or are users blocked on upgrading to 0.9 until they update the raw yaml files.

Thanks for all your hard work!

**Comment by pcrumley** [comment]
Just in case someone finds their way here from googling errors like I did, I wanted to update that i answered my own question by looking at another issue and you can uses singleton maps  https://docs.rs/serde_yaml/0.9.4/serde_yaml/with/singleton_map/index.html

Thanks again for all the hard work!
